File: zad1.py
import math

# Метод Ньютона
def f(x): return 2*x + math.tan(x) # значение ф-ии
def fsh(x): return 2 + 1/(math.cos(x)*math.cos(x)) # значение производной ф-ии
def newton(x0, eps):
    x1 = x0 - f(x0) / fsh(x0)
    iternum = 1
    while abs(x1 - x0) > eps:
        print(iternum, "{:15.6f}".format(x0), "{:15.6f}".format(x1), "{:15.6f}".format(x1 - x0))
        x0 = x1
        x1 = x0 - f(x0) / fsh(x0)
        iternum += 1
    return x1

# Метод простых итераций
def phi(x): return -0.5 * math.tan(x) # итерационная ф-ия
def SimpleIterations(x0, eps):
    x1 = phi(x0)
    iternum = 1
    while abs(x1 - x0) > eps:
        print(iternum, "{:15.6f}".format(x0), "{:15.6f}".format(x1), "{:15.6f}".format(x1 - x0))
        x0 = x1
        x1 = phi(x0)
        iternum += 1
    return x1

def bisection(a, b, eps):
    c = 0.0
    iternum = 1
    while abs(b - a) > eps:
        print(iternum, "{:15.6f}".format(a), "{:15.6f}".format(b), "{:15.6f}".format(b - a))
        c = (a + b) / 2
        if f(a) * f(c) < 0: b = c
        else: a = c
        iternum += 1
    return c


# Метод Ньютона
print("Метод Ньютона:")
print("N".ljust(15) + "xn".ljust(15) + "xn+1".ljust(15) + "xn+1 - xn")
x0 = 1.5  # начальное приближение
eps = 0.001  # точность решения
x = newton(x0, eps)
print("Корень:", x, "\n")

# Метод простых итераций
print("Метод простых итераций:")
print("N".ljust(15) + "xn".ljust(15) + "xn+1".ljust(15) + "xn+1 - xn")
x0 = 1.0  # начальное приближение
eps = 0.001  # точность решения
x = SimpleIterations(x0, eps)
print("Корень:", x, "\n")

# Метод половинного деления
print("Метод половинного деления:")
print("N".ljust(15) + "an".ljust(15) + "bn".ljust(15) + "bn - an")
a = -2.0  # начальный интервал
b = 2.0
eps = 0.001  # точность решения
x = bisection(a, b, eps)
print("Корень:", x, "\n")

File: zad1.cpp
// 2𝑥 + 𝑡𝑔(𝑥) = 0
#include <iostream>
#include <math.h>
#include <iomanip>

using namespace std;

// Метод Ньютона
double f(double x) { // значение функции
    return 2*x + tan(x);
}
double fsh(double x) { // значение производой функции
    return 2 + 1/(cos(x)*cos(x));
}
double newton(double x0, double eps) {
    double x1 = x0 - f(x0) / fsh(x0);
    int iternum = 1;

    while (fabs(x1 - x0) > eps) {
        cout << iternum << setw(15) << x0 << setw(15) << x1 << setw(15) << x1 - x0 << endl;
        x0 = x1;
        x1 = x0 - f(x0) / fsh(x0);
        iternum++;
    }
    return x0;
}

// Метод простых итераций
double phi(double x) { // итерационная функция
    return x - f(x) / fsh(x);
}
double SimpleIterations(double x0, double eps) {
    double x1 = phi(x0);
    int iternum = 1;
    while (fabs(x1 - x0) > eps) {
        cout << iternum << setw(15) << x0 << setw(15) << x1 << setw(15) << x1 - x0 << endl;
        x0 = x1;
        x1 = phi(x0);
        iternum++;
    }
    return x0;
}

// Метод половинного деления
double bisection(double a, double b, double eps) {
    double c = 0.0;
    int iternum = 1;
    while (fabs(b - a) > eps) {
        cout << iternum << setw(15) << a << setw(15) << b << setw(15) << b - a << endl;
        c = (a + b) / 2;
        if (f(c) == 0.0) {
            break;
        } else if (f(c) * f(a) < 0){
            b = c;
        } else {
            a = c;
        }
        iternum++;
    }
    return c;
}


int main() {
    setlocale(LC_ALL, "Rus");
    // Метод Ньютона
    cout << "Метод Ньютона:" << endl << "N" << setw(15) << "xn" << setw(15) << "xn+1" << setw(15) << "xn+1 - xn" << endl;
    double x0 = 1.5; // начальное приближение
    double eps = 0.0001; // точность решения
    double x = newton(x0, eps);
    cout << "Корень: " << x << endl << endl;

    // Метод простых итераций
    cout << "Метод простых итераций:" << endl << "N" << setw(15) << "xn" << setw(15) << "xn+1" << setw(15) << "xn+1 - xn" << endl;
    x0 = 1.5; // начальное приближение
    eps = 0.0001; // точность решение
    x = SimpleIterations(x0, eps);
    cout << "Корень: " << x << endl << endl;

    // Метод половинного деления
     cout << "Метод половинного деления:" << endl << "N" << setw(15) << "an" << setw(15) << "bn" << setw(15) << "bn - an" << endl;
    double a = -10.0; // начальный интервал
    double b = 5.0;
    eps = 0.0001; // точность решения
    x = bisection(a, b, eps);
    cout << "Корень: " << x << endl << endl;

    return 0;
}